### PR TITLE
Codegen updates

### LINKED ISSUE #2866: cmd-line tutorial
https://www.libelektra.org/tutorials/command-line-options should also describe:

- how to use gopts
- how to use command-line options with the high-level API

### PR DESCRIPTION
**Pull request by kodebach**
Re-created from #2805

Contains some updates for the code-generation. For progress see #2680.

Fixes #2866

## Basics

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [x] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

- [x] I added unit tests
- [x] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [x] meta data is updated (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))

## Review

Reviewers will usually check the following:

- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)
- [ ] Documentation is concise and good to read
- [ ] Examples are well chosen and understandable

### CODE REVIEW

**Comment by kodebach** [comment]
@markus2330 Once a7 is online again and Jenkins completes all build jobs, this PR can be merged. (At least from my point of view.)

**Comment by markus2330** [comment]
Thank you!

@bauhaus93: here is the new PR. Please try out the tutorials.

File: src/plugins/specload/README.md [hunk 1]
```diff
@@ -80,28 +80,52 @@ with the original specification. Currently this verification is very restrictive
...
+## Limitations
+
+- The plugin only allows the following modifications right now:
   - add/edit/remove `description` or `opt/help`
   - add/edit `default`
   - add `type`
-
-  However, because the default `resolver` is not compatible with plugins that produce a KeySet without a file present, you can only use
-  `noresolver`. This means that you cannot set any values. (You will get an error about a non-existent file, if you try.)
+- The plugin must be mounted using `noresolver` or another resolver that always calls the storage plugin. To work arround
```
Comment (bauhaus93) [inline src/plugins/specload/README.md]:
```suggestion
- The plugin must be mounted using `noresolver` or another resolver that always calls the storage plugin. To work around
```

**Comment by kodebach** [comment]
jenkins build libelektra please

**Comment by kodebach** [comment]
@markus2330 Any idea why `testscr_check_formatting` fails in `debian-stretch-full-mmap-asan` but not in any other job using the same docker image?

**Comment by sanssecours** [comment]
> Any idea why `testscr_check_formatting` fails in `debian-stretch-full-mmap-asan` but not in any other job using the same docker image?

This seems to be a transient error. As far as I can tell, `git diff --quiet` does report code differences in line 69 of the code below, while it does not detect a difference in line 72. As to why this happens, I have no idea.

https://github.com/ElektraInitiative/libelektra/blob/4e0d346dcf637061c20bcd367e8d7debfa32f25c/tests/shell/check_formatting.sh#L69-L76

Anyway, I opened pull request #2929, which hopefully gets rid of this error.

**Comment by kodebach** [comment]
But then the same thing would occur on all the other build jobs as well, wouldn't it? (At least the ones using the same docker image)

**Comment by sanssecours** [comment]
> But then the same thing would occur on all the other build jobs as well, wouldn't it?  (At least the ones using the same docker image)

Unfortunately, since this is a transient error (it only happens in rare cases), I think the answer to this question is no.

Currently one option to get rid of the error is to restart the build. Since this is a rather bad “solution” – especially considering the stability of the Jenkins build servers  –, I opened PR #2929. At least the new version of the test should provide more information on what the specific problem is.

Anyway, I am sorry for the false positive. I am pretty I wrote the code that causes these problems.

**Comment by kodebach** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
@bauhaus93 do you have any troubles reviewing the PR?

**Comment by kodebach** [comment]
We need an update to the `debian` branch, since I removed `kdbconfig.h` from the installed files. Should I update it first, or do we want to merge the PR before updating `debian`?

**Comment by markus2330** [comment]
You can update the debian branch by directly pushing to it (to remove the one file).

**Comment by kodebach** [comment]
I made the required change to the debian branch.

@markus2330 Is it possible for you (when you are logged in) to restart the Jenkins build from the "Build artifacts" stage? It would be a bit of a waste to repeat all the successful builds, when nothing inside the PR has changed.

**Comment by sanssecours** [comment]
> Is it possible for you (when you are logged in) to restart the Jenkins build from the "Build artifacts" stage?

Unfortunately, [as far as I know that is not possible](https://issues.jenkins-ci.org/browse/JENKINS-45455).

> It would be a bit of a waste to repeat all the successful builds, when nothing inside the PR has changed.

I think so too. I restarted the whole Jenkins build for now.

**Comment by kodebach** [comment]
> Unfortunately, [as far as I know that is not possible](https://issues.jenkins-ci.org/browse/JENKINS-45455).

That issue is marked as "Resolved" and the [release notes](https://plugins.jenkins.io/pipeline-model-definition), mention that it was release as part of version 1.3. I am not sure which version we are using.

**Comment by sanssecours** [comment]
> > Unfortunately, [as far as I know that is not possible](https://issues.jenkins-ci.org/browse/JENKINS-45455).
> 
> That issue is marked as "Resolved" and the [release notes](https://plugins.jenkins.io/pipeline-model-definition), mention that it was release as part of version 1.3. I am not sure which version we are using.

As far as I know we do not use the [declarative syntax for pipeline stages](https://jenkins.io/doc/book/pipeline/syntax/#declarative-pipeline), which means we are out of luck according to the [issue description](https://issues.jenkins-ci.org/browse/JENKINS-45455):

> ## Out of scope
> - Automatically stash/unstash the workspace
> - Scripted pipeline (technical limitation)

.

**Comment by kodebach** [comment]
@markus2330 I suggest we merge this huge PR now, since all build jobs succeeded. If there is anything else that needs to change, we can always create a new PR.

File: cmake/ElektraConfig.cmake.in [hunk 2]
```diff
@@ -2,32 +2,29 @@
 #
 # Variables for the Elektra library:
 #
-#  Elektra_INCLUDE_DIRS  - where to find kdb.h, etc.
-#  Elektra_LIBRARIES     - List of libraries when using Elektra.
+# Elektra_INCLUDE_DIRS  - where to find kdb.h, etc. Elektra_LIBRARIES     - List of libraries when using Elektra.
```
Comment (markus2330) [inline cmake/ElektraConfig.cmake.in]: This formatting got broken.

@sanssecours any idea how to avoid this?

File: .travis.yml [hunk 3]
```diff
@@ -82,7 +82,6 @@ matrix:
           -multifile;\
           -network;\
           -ni;\
-          -noresolver;\
```
Comment (markus2330) [inline .travis.yml]: Why the change?

File: cmake/ElektraConfig.cmake.in [hunk 4]
```diff
@@ -46,3 +43,81 @@ set (Elektra_BUILD_TYPE @Elektra_BUILD_TYPE@)
...
 set (Elektra_STATIC @BUILD_STATIC@)
+
+# ~~~
+# Invoke the `kdb gen` during the build.
+# 
+# elektra_kdb_gen (template parent_key output_name
+#                  INPUT <plugin> <file>
+#                  OUTPUT_DIR <output_dir>
+#                  KDB <kdb>
+#                  PARAMS <params>...
```
Comment (markus2330) [inline cmake/ElektraConfig.cmake.in]: Isn't PARAMS optional?

File: cmake/ElektraConfig.cmake.in [hunk 5]
```diff
@@ -46,3 +43,81 @@ set (Elektra_BUILD_TYPE @Elektra_BUILD_TYPE@)
...
+#                  OPTIONS <options>...)
+#
+# This will call `<kdb> gen -F <plugin>=<file> <options>... template parent_key output_name <params>...` in the
+# directory <output_dir>.
+#
```
Comment (markus2330) [inline cmake/ElektraConfig.cmake.in]: Can you also give an example of how to execute it?

File: doc/THIRD-PARTY-LICENSES [hunk 6]
```diff
@@ -13,3 +13,4 @@ src/tools/website-frontend/resources/assets/skin/base/fonts/font-awesome/font-aw
 src/tools/website-frontend/resources/assets/skin/base/util/space.less: mit
 src/tools/web/client/public/Roboto-Light.ttf: apache-2.0
 src/tools/web/client/public/Roboto-Regular.ttf: apache-2.0
+src/tools/kdb/gen/mustache.hpp: BSL-1.0
```
Comment (markus2330) [inline doc/THIRD-PARTY-LICENSES]: Thank you for maintaining! I made it now mandatory in f256c0dbc8f6c9189aff2e8ae7f20bdfe9c4f16b

File: doc/help/elektra-highlevel-gen.md [hunk 7]
```diff
@@ -0,0 +1,410 @@
+# elektra-highlevel-gen(7) -- High-level API code-generation advanced features
+
+This document focuses on the advanced features of the high-level API code-generator template. We assume you already familiarized yourself
+with the basic features explained in [`kdb-gen-highlevel(1)`](kdb-gen-highlevel.md).
+
+## Configuration Options
+
+The parameters that are relevant to the concepts described here are (for the rest see [`kdb-gen-highlevel(1)](kdb-gen-highlevel.md)):
+
+- `enumConv`: allowed values: `strcmp`, `switch`, `default` (default)
```
Comment (markus2330) [inline doc/help/elektra-highlevel-gen.md]: "auto" instead of "default" might be less confusing.

And reorder these items here so that they match your later explanations.

File: doc/news/_preparation_next_release.md [hunk 8]
```diff
@@ -25,11 +25,16 @@ You can also read the news [on our website](https://www.libelektra.org/news/0.9.
...
 - <<HIGHLIGHT2>>
 - <<HIGHLIGHT3>>
 
-### <<HIGHLIGHT1>>
+### Code Generation
+
+While the new `kdb gen` was already included in the last release, it is now fully functional and ready for use. To get started take a look
+at the new man-page for [`kdb-gen(1)`](https://www.libelektra.org/manpages/kdb-gen). _(Klemens Böswirth)_
+
+If specifically want to use it with the High-Level API take a look at [this tutorial](https://www.libelektra.org/tutorials/high-level-api).
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: sentence?

File: doc/news/_preparation_next_release.md [hunk 9]
```diff
@@ -25,11 +25,16 @@ You can also read the news [on our website](https://www.libelektra.org/news/0.9.
...
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: Also write about cmake integration.

File: doc/news/_preparation_next_release.md [hunk 10]
```diff
@@ -25,11 +25,16 @@ You can also read the news [on our website](https://www.libelektra.org/news/0.9.
...
 
-- <<HIGHLIGHT1>>
+- Code generation
 - <<HIGHLIGHT2>>
 - <<HIGHLIGHT3>>
 
-### <<HIGHLIGHT1>>
+### Code Generation
+
+While the new `kdb gen` was already included in the last release, it is now fully functional and ready for use. To get started take a look
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: productive use?

File: doc/news/_preparation_next_release.md [hunk 11]
```diff
@@ -150,7 +157,7 @@ you up to date with the multi-language support provided by Elektra.
 
 ### CMake
 
-- <<TODO>>
+- `kdbtypes.h` is now generated directly via a CMake `configure_file` call. _(Klemens Böswirth)_
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: Not really important. Better write more about the highlight.

File: doc/tutorials/application-integration.md [hunk 12]
```diff
@@ -12,10 +12,15 @@ In Elektra different forms of application integrations are possible:
    - [Intercept File System](/src/bindings/intercept/fs/README.md)
 3. Integration where applications directly use Elektra to read and
    store settings.
+   1. Using the low-level API.
```
Comment (markus2330) [inline doc/tutorials/application-integration.md]: Is it possible to call it "i" and "ii" here?

File: cmake/ElektraConfig.cmake.in (hunk 2 above)
Comment (sanssecours) [reply cmake/ElektraConfig.cmake.in]: The easiest option is probably using a fence:

```cmake
# ~~~
```

before and after the comment block, just like we did here:

https://github.com/ElektraInitiative/libelektra/blob/bc9254cd9ab37c10275f8e63763d14ee3560bec5/cmake/ElektraCache.cmake#L1-L9

.

File: .travis.yml (hunk 3 above)
Comment (kodebach) [reply .travis.yml]: It is required for `specload` and therefore needed for some of the codegen tests.

File: cmake/ElektraConfig.cmake.in (hunk 4 above)
Comment (kodebach) [reply cmake/ElektraConfig.cmake.in]: Yes, and `OPTIONS`, `KDB` and `OUTPUT_DIR` too. I'll make that more clear.

File: cmake/ElektraConfig.cmake.in (hunk 5 above)
Comment (kodebach) [reply cmake/ElektraConfig.cmake.in]: I can, but does it really help that much, if I add e.g.:

```
elektra_kdb_gen (highlevel "/sw/org/app/#0/current" elektragen ni spec.ini)
```

That is all you need to do. CMake will then know which files are produced by `kdb gen` and call it, if one of the files is required by a target.

File: doc/help/elektra-highlevel-gen.md (hunk 7 above)
Comment (kodebach) [reply doc/help/elektra-highlevel-gen.md]: > And reorder these items here so that they match your later explanations.

Reorder in what way? The order matches exactly with the following paragraphs.

File: doc/tutorials/application-integration.md (hunk 12 above)
Comment (kodebach) [reply doc/tutorials/application-integration.md]: No, markdown only recognises Arabic numbers as list items. GitHub is quite unique in how it renders nested numbered lists. Most other markdown renderers I tried render all levels with Arabic numerals.

I don't know what we use for the website, but I'll change the references in the text below to `3.1` and `3.2`. IMO it is easier to understand that `3.1` references `3.i` than the other way around.

File: .travis.yml (hunk 3 above)
Comment (markus2330) [reply .travis.yml]: Then maybe make it mandatory? (src/plugins/CMakeLists.txt)

File: cmake/ElektraConfig.cmake.in (hunk 5 above)
Comment (markus2330) [reply cmake/ElektraConfig.cmake.in]: I think even this example might help. A full usage example might help even more (how to compile the generated files).

File: doc/tutorials/application-integration.md (hunk 12 above)
Comment (markus2330) [reply doc/tutorials/application-integration.md]: Yes, I agree. Only Arabic numbers is easier.

File: cmake/ElektraConfig.cmake.in (hunk 5 above)
Comment (kodebach) [reply cmake/ElektraConfig.cmake.in]: Full examples can be found in the examples directory.

File: .travis.yml (hunk 3 above)
Comment (kodebach) [reply .travis.yml]: It is not actually mandatory, since `specload` isn't mandatory. Also `specload` can work with other resolvers, it is just that right now `noresolver` is the only one that works. Other resolvers don't call the storage plugin, if no storage file exists.

File: cmake/ElektraConfig.cmake.in (hunk 5 above)
Comment (markus2330) [reply cmake/ElektraConfig.cmake.in]: Then adding links to the examples directories is all what is needed.

File: doc/help/elektra-highlevel-gen.md (hunk 7 above)
Comment (markus2330) [reply doc/help/elektra-highlevel-gen.md]: Forward references like "The `enumConv` option is described [below](#enums)." should be avoided. Moving enumConv the last of these 3 entries would solve that problem (the sentence could be removed).

File: cmake/ElektraConfig.cmake.in (hunk 5 above)
Comment (kodebach) [reply cmake/ElektraConfig.cmake.in]: How should I link to the examples? With a Github link? I don't think the external examples show up on the homepage.

Comment (markus2330) [reply cmake/ElektraConfig.cmake.in]: The website automatically turns all internal links that are not in the website to github links.
